### PR TITLE
Litex SoC minitest (sources)

### PR DESCRIPTION
**Pull request by mkurc-ant**
This test contains LiteX SoC design which utilizes VexRiscV CPU, DDR memory and Ethernet. The design can be implemented using both Yosys+Vivado and Vivado only flow. Output bitstream(s) can then processed using bit2fasm utility which allows to identify missing segbits.

Source code was generated with the following command: `arty.py --cpu-type vexriscv --cpu-variant linux --with-ethernet --no-compile-software --no-compile-gateware`

### CODE REVIEW

**Comment by mithro** [comment]
@mkurc-ant What does the result look like?

**Comment by mkurc-ant** [comment]
@mitrho For Vivado we have:
```
# In frame 0x00000026 4 bits were not converted.
# In frame 0x00000027 2 bits were not converted.
# In frame 0x0040049a 104 bits were not converted.
# In frame 0x0040049b 159 bits were not converted.
# In frame 0x0000151c 93 bits were not converted.
# In frame 0x0000151d 76 bits were not converted.
# In frame 0x0040159a 49 bits were not converted.
# In frame 0x0040159b 50 bits were not converted.
# In frame 0x0040159c 180 bits were not converted.
# In frame 0x0040159d 162 bits were not converted.
# In frame 0x0040159e 246 bits were not converted.
# In frame 0x0040159f 269 bits were not converted.
# In frame 0x004015a0 262 bits were not converted.
# In frame 0x004015a1 266 bits were not converted.
# In frame 0x004015a2 56 bits were not converted.
# In frame 0x004015a3 56 bits were not converted.
# In frame 0x004015a5 2 bits were not converted.
# In frame 0x004015a6 399 bits were not converted.
# In frame 0x004015a7 415 bits were not converted.
```

And for Yosys + Vivado we have:
```
# In frame 0x00000026 3 bits were not converted.
# In frame 0x0040159d 160 bits were not converted.
# In frame 0x0040159e 246 bits were not converted.
# In frame 0x0040159f 268 bits were not converted.
# In frame 0x004015a3 56 bits were not converted.
# In frame 0x0000151c 93 bits were not converted.
# In frame 0x0000151d 76 bits were not converted.
# In frame 0x0040159a 50 bits were not converted.
# In frame 0x0040159b 50 bits were not converted.
# In frame 0x0040159c 182 bits were not converted.
# In frame 0x004015a0 262 bits were not converted.
# In frame 0x004015a1 266 bits were not converted.
# In frame 0x004015a2 56 bits were not converted.
# In frame 0x004015a5 2 bits were not converted.
# In frame 0x004015a6 399 bits were not converted.
# In frame 0x004015a7 415 bits were not converted.
```

It is worth looking for resource utilization logs from Vivado. They contain tile names. It should be easy to spot tiles that we do not have bits for.

**Comment by mithro** [comment]
Is there an easy way to work out what is at each frame address?

**Comment by mithro** [comment]
@mkurc-ant - I ran make in `minitests/litex/src.yosys` and didn't get any output?
```
(env) tansell@tansell:~/github/SymbiFlow/prjxray/minitests/litex/src.yosys$ make
PYTHONPATH="/usr/local/google/home/tansell/github/SymbiFlow/prjxray:/usr/local/google/home/tansell/github/SymbiFlow/prjxray/utils:/usr/local/google/home/tansell/github/SymbiFlow/prjxray/third_party/fasm" \     
    PATH="/usr/local/google/home/tansell/github/SymbiFlow/prjxray/build/tools:/usr/local/google/home/tansell/github/SymbiFlow/prjxray/env/bin:/usr/local/google/home/tansell/bin:/usr/local/google/home/tansell/gocode/bin:/usr/local/google/home/tansell/go/bin:/usr/local/google/home/tansell/.nvm/versions/node/v8.8.1/bin:/usr/local/google/home/tansell/bin:/usr/local/google/home/tansell/gocode/bin:/usr/local/google/home/tansell/go/bin:/usr/lib/google-golang/bin:/usr/local/buildtools/java/jdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" \                                                                          
    python3 \
        /usr/local/google/home/tansell/github/SymbiFlow/prjxray/utils/bit2fasm.py \
        --db-root "/usr/local/google/home/tansell/github/SymbiFlow/prjxray/database/artix7" --part "/usr/local/google/home/tansell/github/SymbiFlow/prjxray/database/artix7/xc7a35tcsg324-1" --verbose \          
        top.bit >top.fasm \
        || (rm -f top.fasm && exit -1)
(env) tansell@tansell:~/github/SymbiFlow/prjxray/minitests/litex/src.yosys$
```

**Comment by mkurc-ant** [comment]
@mithro I updated the Vivado invocation in Makefiles so it now uses `utils/vivado.sh` script.

The Makefile for Yosys+Vivado flow does the following:
- Builds Yosys if needed (its a submodule under `third_party/yosys`)
- Generates `top.edif` using Yosys synthesis flow
- Invokes Vivado in order to generate bitstream (`top.bit`)
- Invokes the `bit2fasm.py` to generate `top.fasm` file.

Since the `bit2fasm.py` outputs the FASM to stdout its outpus is redirected to the FASM file. If an error occurs, the FASM file would contain the error message instead of FASM feature list. So in case of `bit2fasm.py` failure the output file is deleted and the Makefile shoud return -1 exitcode.

Can you verify that all the above steps completed successfully and that you have all the output files generated ?

**Comment by tmichalak** [comment]
@mithro 
> Is there an easy way to work out what is at each frame address?

Segprint is capable of decoding the frames into specific tiles. It also tells which bits are not in the database at all. e.g. for litex with vivado frontend:
```
$XRAY_SEGPRINT -z -D -b top.bits > segprint.log
head -30 segprint.log
Non-database bits:
bit_0040159a_050_23
bit_0040159b_050_14
bit_0040159b_050_15
bit_0040159c_050_22
bit_0040159c_050_23
bit_0040159c_050_24
bit_0040159d_050_17
bit_0040159d_050_19
bit_0040159d_050_22
bit_0040159d_050_30
bit_0040159d_050_31
bit_0040159e_050_15
bit_0040159e_050_16
bit_0040159e_050_19
bit_0040159f_050_21
bit_0040159f_050_30
bit_004015a0_050_14
bit_004015a5_050_28
bit_004015a5_050_29
bit_004015a6_050_15
bit_004015a6_050_26
bit_004015a7_050_14
bit_004015a7_050_15
bit_004015a7_050_16
bit_004015a7_050_22
```
We could integrate it into the Makefile.

**Comment by mithro** [comment]
@tmichalak How do I get a `.bits` file?

**Comment by tmichalak** [comment]
@mithro
With bitread, once you source prjxray virualenv you can call:
${XRAY_BITREAD} -F ${XRAY_ROI_FRAMES} -o top.bits -z -y top.bit

**Comment by mkurc-ant** [comment]
I updated the code for LiteX SoC so now it is Linux compatible.

File: minitests/litex/src.vivado/Makefile [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+.PHONY: all clean
+
+top.bit: $(VIVADO)
+	mkdir -p build
+	cd build && $(VIVADO) -mode batch -source ../top.tcl -nojournal -tempDir build -log vivado.log -verbose
+	cp build/*.bit ./
+
+top.fasm: top.bit
+	PYTHONPATH="$(BASE_PATH):$(BASE_PATH)/utils:$(BASE_PATH)/third_party/fasm" \
```
Comment (litghost) [inline minitests/litex/src.vivado/Makefile]: Just use ``$(XRAY_BIT2FASM)`` over ``$(BASE_PATH)/utils/bit2fasm.py``

File: minitests/litex/src.vivado/Makefile [hunk 2]
```diff
@@ -0,0 +1,35 @@
+BASE_PATH = $(abspath $(PWD)/../../../)
+VIVADO = $(BASE_PATH)/utils/vivado.sh
```
Comment (litghost) [inline minitests/litex/src.vivado/Makefile]: Just use ``$(XRAY_VIVADO)``

File: minitests/litex/src.vivado/Makefile [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+PART = xc7a35tcsg324-1
+BIT2FASM_ARGS = --db-root "$(BASE_PATH)/database/artix7" --part "$(BASE_PATH)/database/artix7/$(PART)" --verbose
```
Comment (litghost) [inline minitests/litex/src.vivado/Makefile]: ``-db-root`` should not be required if ``prjxray/settings/artix.sh`` was sourced.

**Comment by mkurc-ant** [comment]
@litghost I fixed makefile stuff

File: minitests/litex/src.vivado/Makefile [hunk 4]
```diff
@@ -0,0 +1,33 @@
+BASE_PATH = $(abspath $(PWD)/../../../)
```
Comment (litghost) [inline minitests/litex/src.vivado/Makefile]: Use ``${XRAY_DIR}``?

File: minitests/litex/src.vivado/Makefile [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	@rm -f *.bit
+	@rm -f *.bin
+	@rm -f *.bits
+	@rm -f *.fasm
+	@rm -f *.log
+	@rm -rf build
+
+.PHONY: all clean
+
+top.bit: $(VIVADO)
```
Comment (litghost) [inline minitests/litex/src.vivado/Makefile]: Dependency seems wrong?  Should be top.tcl and maybe *.v?

**Comment by mkurc-ant** [comment]
@litghost I added those dependencies you requested.

**Pull request merged**